File: plugins/probe-rat-layer/probe_layer/core/metrics.py
"""分层质量指标：软冲突 + 数量/长短/扇区均衡。

所有指标都是"越小越好"。扇区均衡只统计**实际有线经过**的扇区：
圆形针卡测试机端不满配，部分角度没有外围连线，空扇区不应被惩罚。
圆心固定为 (0,0)（用户确认）。
"""
from __future__ import annotations

from collections import Counter


def sector_index(angle_deg: float, sector_angle_deg: float) -> int:
    """极角 [0,360) → 扇区号 [0, n)。n = round(360 / sector_angle_deg)。"""
    n = max(1, int(round(360.0 / sector_angle_deg)))
    return int(angle_deg // sector_angle_deg) % n


def soft_crossings(assignment: dict[str, int],
                   soft_pairs: list[tuple[str, str]]) -> int:
    """同层软冲突（交叉）对数。只统计两端都已分配的线对。"""
    total = 0
    for a, b in soft_pairs:
        la, lb = assignment.get(a), assignment.get(b)
        if la is not None and la == lb:
            total += 1
    return total


def count_imbalance(layer_counts: dict[int, int]) -> float:
    """各层线数不均衡度 = (max - min) / mean，0 表示完全均衡。

    空层（0 线）也算一层：全堆在一层时是不均衡的。
    """
    vals = list(layer_counts.values())
    if len(vals) < 2:
        return 0.0
    mean = sum(vals) / len(vals)
    if mean <= 0:
        return 0.0
    return (max(vals) - min(vals)) / mean


def length_imbalance(layer_lengths: dict[int, float],
                     layer_counts: dict[int, int]) -> float:
    """各层平均线长不均衡度 = (max_avg - min_avg) / mean_avg。

    空层平均线长按 0 计（一个空层即失衡）。
    """
    avgs = [layer_lengths[l] / c if c > 0 else 0.0
            for l, c in layer_counts.items()]
    if len(avgs) < 2:
        return 0.0
    mean = sum(avgs) / len(avgs)
    if mean <= 0:
        return 0.0
    return (max(avgs) - min(avgs)) / mean


def sector_imbalance(layer_sectors: dict[int, Counter], total_wires: int) -> float:
    """各层扇区覆盖不均衡度，只统计实际有线经过的扇区。

    对每个"有线的扇区" s，取它落在各层的线数 c_s 的 (max-min) 求和，
    再除以总线数归一化到 [0,1) 附近。空扇区不参与，避免无意义惩罚。
    """
    populated: set[int] = set()
    for c in layer_sectors.values():
        populated.update(c.keys())
    if not populated or total_wires <= 0:
        return 0.0
    imbalance = 0.0
    for s in populated:
        vals = [layer_sectors[l].get(s, 0) for l in layer_sectors]
        imbalance += max(vals) - min(vals)
    return imbalance / total_wires

File: core-plugins/probe-rat-layer/src/metrics.rs
//! 分层质量指标：软冲突 + 数量/长短/扇区均衡（移植自 Python `core/metrics.py`）。

use crate::collections::HashMap;

/// 极角 [0,360) → 扇区号 [0, n)。n = round(360 / sector_angle_deg)。
pub fn sector_index(angle_deg: f64, sector_angle_deg: f64) -> i64 {
    let n = (360.0 / sector_angle_deg).round().max(1.0) as i64;
    let a = if angle_deg < 0.0 { angle_deg + 360.0 } else { angle_deg };
    (((a / sector_angle_deg) as i64) % n + n) % n
}

/// 同层软冲突（交叉）对数。只统计两端都已分配的线对。
pub fn soft_crossings(assignment: &HashMap<String, i64>, soft_pairs: &[(String, String)]) -> i64 {
    let mut total = 0;
    for (a, b) in soft_pairs {
        if let (Some(la), Some(lb)) = (assignment.get(a), assignment.get(b)) {
            if la == lb {
                total += 1;
            }
        }
    }
    total
}

/// 各层线数不均衡度 = (max - min) / mean，0 表示完全均衡。空层也算一层。
pub fn count_imbalance(layer_counts: &HashMap<i64, i64>) -> f64 {
    let vals: Vec<i64> = layer_counts.values().copied().collect();
    if vals.len() < 2 {
        return 0.0;
    }
    let sum: i64 = vals.iter().sum();
    let mean = sum as f64 / vals.len() as f64;
    if mean <= 0.0 {
        return 0.0;
    }
    let max = *vals.iter().max().unwrap() as f64;
    let min = *vals.iter().min().unwrap() as f64;
    (max - min) / mean
}

/// 各层平均线长不均衡度 = (max_avg - min_avg) / mean_avg。空层平均线长按 0 计。
pub fn length_imbalance(layer_lengths: &HashMap<i64, f64>, layer_counts: &HashMap<i64, i64>) -> f64 {
    let keys: Vec<&i64> = layer_counts.keys().collect();
    if keys.len() < 2 {
        return 0.0;
    }
    let avgs: Vec<f64> = keys
        .iter()
        .map(|l| {
            let c = *layer_counts.get(*l).unwrap_or(&0);
            if c > 0 {
                layer_lengths.get(*l).copied().unwrap_or(0.0) / c as f64
            } else {
                0.0
            }
        })
        .collect();
    let mean = avgs.iter().sum::<f64>() / avgs.len() as f64;
    if mean <= 0.0 {
        return 0.0;
    }
    let max = avgs.iter().cloned().fold(f64::NEG_INFINITY, f64::max);
    let min = avgs.iter().cloned().fold(f64::INFINITY, f64::min);
    (max - min) / mean
}

/// 各层扇区覆盖不均衡度，只统计实际有线经过的扇区。
pub fn sector_imbalance(layer_sectors: &HashMap<i64, HashMap<i64, i64>>, total_wires: i64) -> f64 {
    let mut populated: crate::collections::HashSet<i64> = crate::collections::HashSet::default();
    for c in layer_sectors.values() {
        for k in c.keys() {
            populated.insert(*k);
        }
    }
    if populated.is_empty() || total_wires <= 0 {
        return 0.0;
    }
    let mut imbalance = 0.0;
    for s in populated {
        let mut maxv = 0i64;
        let mut minv = i64::MAX;
        for l in layer_sectors.keys() {
            let v = layer_sectors[l].get(&s).copied().unwrap_or(0);
            maxv = maxv.max(v);
            minv = minv.min(v);
        }
        if minv == i64::MAX {
            minv = 0;
        }
        imbalance += (maxv - minv) as f64;
    }
    imbalance / total_wires as f64
}
